### PR TITLE
Allow `joins` to be unscoped

### LINKED ISSUE #13775: Model.unscoped block not removing default scope for joins
Summary:

Suppose Model has a default scope.

```ruby
Model.unscoped { Foo.joins(:model) }
```

acts as if `Model.unscoped { ... }` were not specified. That is, it constructs `Foo.joins(:model)` as if it weren't inside `Model.unscoped`, so that `Model.default_scope` is mistakenly continuing to apply to the `joins`, even though `unscoped` should have removed it.

----

Consider a `User` with a `default_scope` that specifies `status = 1`:

```ruby
>> User.all.to_sql
#=> "SELECT `users`.* FROM `users`  WHERE `users`.`status` = 1"
```

If an `Item` belongs to `User`, trying to join the two with `Item.joins(:user)` seems to always respect `User.default_scope`, no matter how you try to remove the default scope. So the `joins` SQL

```ruby
>> Item.joins(:user).to_sql # includes User.default_scope, CORRECTLY.
#=> "SELECT `items`.* FROM `items` INNER JOIN `users` ON `users`.`id` = `items`.`user_id` AND `users`.`status` = 1
```

correctly includes

```
`users`.`status` = 1
```

as it should, but if you try to remove it with `User.unscoped { ... }`, the `status = 1` default scope on `User` is still applying, which I believe is a bug:

```ruby
User.unscoped do
  Item.joins(:user) # User is incorrectly still scoped to status == 1 here
end
```

As a total aside, it would be nice to have something like `joins_unscoped(:user)` as an alternative.

### PR DESCRIPTION
**Pull request by k0kubun**
Closes https://github.com/rails/rails/issues/13775

Because `unscoped` alters current_scope, join_dependency should use it if exists.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
This need a rebase. Could you check it?

**Comment by k0kubun** [comment]
@rafaelfranca I rebased and pushed.

**Comment by k0kubun** [comment]
@rafaelfranca I rebased and pushed. Could you review this PR?

**Comment by ppworks** [comment]
:+1:

File: activerecord/test/cases/scoping/default_scoping_test.rb [hunk 1]
```diff
@@ -361,6 +361,10 @@ def test_unscoped_with_named_scope_should_not_have_default_scope
     assert_equal 10, DeveloperCalledJamis.unscoped { DeveloperCalledJamis.poor }.length
   end
 
+  def test_unscoped_with_joins_should_not_have_default_scope
+    assert_equal FirstPost.unscoped { Comment.joins(:first_post).to_a }, Comment.joins(:post).to_a
```
Comment (rafaelfranca) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: This test is hard to follow. Why the unscoped version of `Comment.joins(:first_post).to_a` is equals to `Comment.joins(:post).to_a`. Should not the join be removed?

Comment (k0kubun) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: I wanted to join a model with default_scope to unscope.
Because `FirstPost` has [default_scope](https://github.com/rails/rails/blob/bdeeca84e34d3e0863a6fd03defb9c95fd0241d9/activerecord/test/models/post.rb#L189) and `Post` is its unscoped version, I choosed it.

> This test is hard to follow.

I see. `:first_post` may not be a good name for this test case. I should use `:post_with_default_scope`.
I'll change so.

> Why the unscoped version of Comment.joins(:first_post).to_a is equals to Comment.joins(:post).to_a. Should not the join be removed?

That's because the unscope target is not Comment but FirstPost.

Comment (k0kubun) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: I changed the test in ce343bc.

**Comment by Envek** [comment]
@k0kubun thank you very much for this PR! Strongly need it to be merged :smiley: 

I've adapted it to 4.2.0 here: https://github.com/Envek/rails/commit/59944a3d85fe3f1586e85001b2bc071084140a5f
If I create a PR from it to 4-2-stable branch (with all credits to @k0kubun) will it be accepted?

**Comment by k0kubun** [comment]
@Envek Thank you for your feedback for this PR. :smile: I want this to be merged too.

> If I create a PR from it to 4-2-stable branch (with all credits to @k0kubun) will it be accepted?

No problem. But maybe it will not be merged before this PR is accepted.

File: activerecord/lib/active_record/associations/join_dependency/join_association.rb [hunk 2]
```diff
@@ -54,12 +54,18 @@ def join_constraints(foreign_table, foreign_klass, node, join_type, tables, scop
...
+              if klass.current_scope
+                klass.current_scope.clone
+              else
+                relation = ActiveRecord::Relation.create(
+                  klass,
+                  table,
+                  predicate_builder,
+                )
+                klass.send(:build_default_scope, relation)
+              end
```
Comment (sgrif) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: I think this entire expression can be replaced with `klass.all.spawn`

Comment (arthurnn) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: I was looking into that too.. `klass.all` will also respect the default_scope, so that might be worth trying.

Comment (k0kubun) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: That's cool. It passes the test I added. I can't remember why I wrote such a complicated code.
Fixed 8cf0dce

Comment (k0kubun) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: Hmm... other tests were borken.

**Comment by arthurnn** [comment]
Seems like the last commit broke the build, I think using `.all` wont work, as if there is no scope it will :bomb:

**Comment by k0kubun** [comment]
> Seems like the last commit broke the build, I think using .all wont work, as if there is no scope it will :bomb:

Thank you for your comment. I cancelled the commit.

**Comment by sgrif** [comment]
Let me take one more look at this today and then I'll merge

File: activerecord/lib/active_record/associations/join_dependency/join_association.rb (hunk 2 above)
Comment (arthurnn) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: You can probably use `scope_attributes?` to check if there is a scope associated, see https://github.com/rails/rails/pull/19301

Comment (k0kubun) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: Ah, that seems a nice helper for this. I'll try it and run tests.

Comment (k0kubun) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: I tried some code using it, but all of them broke the tests.
Though I don't still look into it well, I have no idea to make use of `scope_attributes?`. `scope_attributes?` doesn't distinguish the case to return current_scope and one to return the scope in L61-66, does it?

Comment (k0kubun) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: Any attempts to refactor the current code crashes the test in 70a5e5638bfd3b2543bb9d4805cbefec84b6b1ce.

**Comment by arthurnn** [comment]
@k0kubun thanks for working on this.
Can you also add a test to make sure `joins` will work with default_scope, without the unscoped block.

**Comment by k0kubun** [comment]
> @k0kubun thanks for working on this.

Thank you for your helpful comments too.

> Can you also add a test to make sure joins will work with default_scope, without the unscoped block.

By your comment, I've just noticed my previous test didn't work because of the change in ce343bc62eeec69b22aa624cd3fefdf026268933.
I fixed the test, added test_default_scope_with_joins and squashed the commits.

File: activerecord/test/cases/scoping/default_scoping_test.rb [hunk 3]
```diff
@@ -361,6 +361,16 @@ def test_unscoped_with_named_scope_should_not_have_default_scope
     assert_equal 10, DeveloperCalledJamis.unscoped { DeveloperCalledJamis.poor }.length
   end
 
+  def test_default_scope_with_joins
+    assert_equal 4,  Comment.joins(:special_post_with_default_scope).length
+    assert_equal 11, Comment.joins(:post).length
```
Comment (arthurnn) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: sorry, but where did you get this numbers from?

Comment (k0kubun) [inline activerecord/test/cases/scoping/default_scoping_test.rb]: That seemed a magick number. I wrote code to count them. 16faa5f85fdee62723bab8f428d7a1455e2bd00f

**Comment by arthurnn** [comment]
Seems like the `unscoped` case is working with this patch, however the `default_scope` still wont work. You can use https://gist.github.com/pallan/8542994 to test that.
Anyhow, I think it is fine for us to fix only the `unscoped` case in this PR, and open another one to fix the `default_scope` case. Not 100% sure about that tho.
@rafaelfranca WDYT?
thanks

**Comment by rafaelfranca** [comment]
I'd fix both cases since for me doesn't make sense to not fix the default_scope case, and we will introduce inconsistency in our codebase

**Comment by k0kubun** [comment]
> Seems like the unscoped case is working with this patch, however the default_scope still wont work. 

> I'd fix both cases since for me doesn't make sense to not fix the default_scope case, and we will introduce inconsistency in our codebase

I guess you think there are 2 problems to fix described in https://github.com/rails/rails/issues/13775.

- `Bar.unscoped { Foo.joins(:bar_with_default_scope) }` should not include bar's default_scope.
- `Foo.joins(:bar_with_default_scope)` should not include bar's default_scope.

But I think the latter case is not a problem because `joins`'s arguments is not a table, which does not have a default_scope, but a underscored model, which has a default_scope. And `unscoped` is a method to remove default_scope from a model. There is no inconsistency for me.

I thought `Foo.joins(:bar_with_defualt_scope)` should include default_scope like now and didn't touch the behavior. So I don't want https://gist.github.com/pallan/8542994 to be passed.

I believe the author of https://github.com/rails/rails/issues/13775 says that not-working unscoped is a bug and does not say "`Item.joins(:user)` seems to always respect `User.default_scope`" is a bug.

**Comment by k0kubun** [comment]
@aprescott What do you think for this?

**Comment by aprescott** [comment]
> I believe the author of #13775 says that not-working unscoped is a bug and does not say "`Item.joins(:user)` seems to always respect `User.default_scope`" is a bug.

As the author in question, this is correct. I did not intend to say that `Item.joins(:user)` should ignore `User.default_scope`. That is, `Item.joins(:user)` should not act identically to `User.unscoped { Item.joins(:user) }`. Only that `User.unscoped { Item.joins(:user) }` is ignoring the `User.unscoped { ... }` block.

> But I think the latter case is not a problem because `joins`'s arguments is not a table, which does not have a default_scope, but a underscored model, which has a default_scope. And `unscoped` is a method to remove default_scope from a model. There is no inconsistency for me.

I would agree with this argument, I think. Removing a default scope for _all_ `joins` calls, independent of any `unscoped` blocks, would feel a little broken to me, as a user.

**Comment by aprescott** [comment]
I will amend #13775 to try and remove this ambiguity, if that'll help.

**Comment by aprescott** [comment]
Updated. Again: `Item.joins(:user)` including `User.default_scope` is correct to me (and would be broken otherwise). The bug here is `User.unscoped { Item.joins(:user) }` acting exactly like `Item.joins(:user)`, as if `User.unscoped { ... }` weren't there.

**Comment by k0kubun** [comment]
I'm glad to hear that. Thank you for your comment.

@sgrif @arthurnn @rafaelfranca What should I update for this?

File: activerecord/lib/active_record/associations/join_dependency/join_association.rb (hunk 2 above)
Comment (sidot3291) [inline activerecord/lib/active_record/associations/join_dependency/join_association.rb]: Woops - I'm still figuring out github and just worked on the same issue today.  @sgrif is right in thinking that `klass.all.spawn` will return the relation with default_scope applied if it hasn't been scoped by a block.   The issue is that `klass.all.spawn.table` is not necessarily the Arel::Table you should be using, because it doesn't have the join alias associated with it.  This is what I came up with as a fix: https://github.com/rails/rails/pull/20937

**Comment by mockdeep** [comment]
Bump on this. We're having to work around this in our codebase with method overrides and tricky sql.

**Comment by k0kubun** [comment]
Rebased and passing tests on travis. 

@rafaelfranca Could you review this?

**Comment by tawan** [comment]
I also took a look the PR and composed a [summary](https://github.com/rails/rails/issues/13775#issuecomment-177249597) in the original issue.
IMO this PR solves the problem quite elegant.

**Comment by jschroeder9000** [comment]
It would be really great if this could be merged before it loses momentum again.  Currently I have to monkey patch ActiveRecord to get expected behavior.

**Comment by jschroeder9000** [comment]
This just made my day.  Thank you to all involved!  :heart:

**Comment by kenn** [comment]
Any chance of backporting this to 4.2-branch?

I also wanted to point out that in case of soft delete, `belongs_to` associations almost always come with `unscope`, like so:

```ruby
class Item < ActiveRecord::Base
  belongs_to :user, -> { unscope(where: :deleted) }
end
```

and in this case, `Item.joins(:user)` **should** behave exactly like `User.unscoped { Item.joins(:user) }`, because the intention is already expressed in the `belongs_to` association.

Is it covered already, or should it go to a new ticket?

**Comment by krzysiek1507** [comment]
If code below is correct, then calling `current_scope` in 
```ruby
Test.joins(other_tests: :test).test_it
```
causes `#<SystemStackError: stack level too deep>` (Rails 5.0.0)

```ruby
class Test < ActiveRecord::Base
  has_many :other_tests

  def self.test_it
    # call here: current_scope
  end
end

class OtherTest < ActiveRecord::Base
  belongs_to :test
end
```

**Comment by KevinSjoberg** [comment]
Is this really fixed? I'm pointing my gem to the HEAD of the Rails `4-2-stable` branch. I've confirmed that the patch is there using `bundle open activerecord`.

I have a default scope on my product `default_scope { where(deleted_at: nil) }` and running `Product.unscoped { Booking.joins(:product) }` still includes the default scope.

**Comment by k0kubun** [comment]
How about `Product.unscoped { Booking.joins(:product).to_a }`?

**Comment by KevinSjoberg** [comment]
@k0kubun that actually worked! How come?

**Comment by k0kubun** [comment]
`Product.unscoped { ... }` affects queries executed inside the block. `Booking.joins(:product)` does not execute a query and just return `ActiveRecord::Relation`, whose query is not determined and able to be modified.

To execute a query, you should call `Booking.joins(:product).first`, `Booking.joins(:product).to_a`, `Booking.joins(:product).find(...)` or something.

**Comment by aprescott** [comment]
It's not always possible/feasible to have the block given to `unscoped` force the query to be executed. It's sometimes necessary to build up a relation, later use that relation to build it up further, and only _then_ execute it.

```ruby
base = Product.unscoped { ... }
base.where(foo: bar)
```

If `unscoped` is still ignored in that case, then yeah I'd say it should be considered broken.

**Comment by k0kubun** [comment]
> It's not always possible/feasible to have the block given to unscoped force the query to be executed. 

I agree with that point.

> It's sometimes necessary to build up a relation, later use that relation to build it up further, and only then execute it.
> base = Product.unscoped { ... }
> base.where(foo: bar)
> If unscoped is still ignored in that case, then yeah I'd say it should be considered broken.

As far as I know, ActiveRecord does not provide such a feature. It's not broken and your proposal is just a feature request.

**Comment by kenn** [comment]
I have a similar problem, which turns out to stem from the same root cause: https://github.com/rails/rails/issues/20679

From the use case standpoint, lack of ability to undo joins makes the `default_scope` feature less usable for soft-delete on anything more than very simple app. Because we use `joins` to exclude soft-deleted associations, but sometimes we need to undo that. (e.g. for admins)

In the above ticket, I've introduced a new `joins_unscoped` method.

```ruby
Booking.joins(:product).joins_unscoped(:product)
```

But from the API standpoint, It may be better to use `unscope` like so?

```ruby
Booking.joins(:product).unscope(joins: :product)
```

**Comment by k0kubun** [comment]
For my personal use case that encouraged to make this patch, I use `unscoped` to disable soft-delete on admin page, which is used for real operation in my company.

In that use case, we could use `around_action` to wrap all queries in controllers for admin page. I think it's quite straight-forward and not confusing. You can use this for action-level usage too. While I don't understand your use case well, I admit that `unscoped` can't help you if you want to remove soft-delete in some limited parts in action.

**Comment by cars10** [comment]
Any chance to see this PR merged into the older rails versions? (4.1.X, 4.2.X)

**Comment by k0kubun** [comment]
> Any chance to see this PR merged into the older rails versions? (4.1.X, 4.2.X)

It's already backported to 4.2.X https://github.com/rails/rails/pull/25232

**Comment by santervo** [comment]
@k0kubun 

> As far as I know, ActiveRecord does not provide such a feature. It's not broken and your proposal is just a feature request.

In my opinion Rails does provide such feature, but not for joins, so it's inconsistent at least.

Here query without join (default scope removed):
```
Post.unscoped { Comment.where(id: 1) }.to_sql
# SELECT "comments".* FROM "comments" WHERE "comments"."id" = $1
```

And here one with join (default scope is not removed)
```
Post.unscoped { Comment.joins(:post) }.to_sql
# SELECT "comments".* FROM "comments" INNER JOIN "posts" ON "posts"."id" = "comments"."post_id" AND "posts"."deleted_at" IS NULL
```

I think this was changed in Rails 4.1.  On Rails 3.2 and 4.0 both examples had default scope removed.

**Comment by k0kubun** [comment]
That's because you misunderstand `unscoped`. You should execute query inside unscoped block but you actually return relation from unscoped block and execute query outside the block.

```rb
irb(main):002:0> Post.unscoped { Comment.where(id: 1).to_sql }
=> "SELECT `comments`.* FROM `comments` WHERE `comments`.`id` = 1"
irb(main):003:0> Post.unscoped { Comment.joins(:post).to_sql }
=> "SELECT `comments`.* FROM `comments` INNER JOIN `posts` ON `posts`.`id` = `comments`.`post_id`"
irb(main):004:0> Comment.joins(:post).to_sql
=> "SELECT `comments`.* FROM `comments` INNER JOIN `posts` ON `posts`.`id` = `comments`.`post_id` AND `posts`.`deleted_at` IS NULL"
irb(main):005:0> Post.unscoped { Comment.joins(:post) }.to_sql
=> "SELECT `comments`.* FROM `comments` INNER JOIN `posts` ON `posts`.`id` = `comments`.`post_id` AND `posts`.`deleted_at` IS NULL"
```

**Comment by k0kubun** [comment]
And please specify the definitions of your models...

I guseed:

```rb
ActiveRecord::Schema.define(version: 20170102230434) do
  create_table "comments", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
    t.integer  "post_id"
    t.datetime "deleted_at"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "posts", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
    t.datetime "deleted_at"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end
end
```

```rb
class Post < ApplicationRecord
  has_many :comments
  default_scope { where(deleted_at: nil) }
end

class Comment < ApplicationRecord
  belongs_to :post
end
```

**Comment by k0kubun** [comment]
And what you want should be not unscoped but unscope. While I don't know old behavior of Rails, Ruby's block should end its effect in the end of the block.

**Comment by edwardmp** [comment]
Is this bug really resolved? We're still seeing a related issue here:
https://github.com/rubysherpas/paranoia/issues/355

**Comment by k0kubun** [comment]
You mention to `unscope` in that issue. `unscope` and `unscope*d*` are totally different things.

**Comment by edwardmp** [comment]
@k0kubun, thanks your right. I'll need to open a separate issue for that.

**Comment by pjmartorell** [comment]
> How about `Product.unscoped { Booking.joins(:product).to_a }`?

Thanks for your suggestion 👍  anyway I think it would be better to use `load` (https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-load) instead of `to_a`:
`Product.unscoped { Booking.joins(:product).load }`

`to_a` returns an `Array` of records while `load` returns an `ActiveRecord::Relation`
